### PR TITLE
Bug 1909587: openstack: Create Control plane nodes sequentially

### PR DESCRIPTION
**Pull request by pierreprinetti**
This patch addresses the OpenStack issue where the Compute module won't
honour the soft-anti-affinity policy when servers are created
concurrently.

### CODE REVIEW

**Comment by pierreprinetti** [comment]
/title WIP: Bug 1909587: openstack: Create Control plane nodes sequentially

This does not apparently work.

**Review by mdbooth — APPROVED** [review]
/lgtm

File: data/data/openstack/masters/main.tf [hunk 1]
```diff
@@ -40,21 +40,84 @@ resource "openstack_compute_servergroup_v2" "master_group" {
...
 
-resource "openstack_compute_instance_v2" "master_conf" {
-  name = "${var.cluster_id}-master-${count.index}"
-  count = var.instance_count
+# The master servers are created in three separate resource definition blocks,
+# rather than a single block with a "count" meta-property, because they need to
+# be created sequentially rather than concurrently by Terraform.
+#
+# The reason why they need to be created one at a time is that OpenStack's
+# Compute module is currently unable to honour the "soft-anti-affinity" policy
```
[the remark this answers is not in the document]
Comment (pierreprinetti) [reply data/data/openstack/masters/main.tf:48]: HM please

**Comment by mandre** [comment]
At first glance this looks OK because we require a minimum of 3 control plane nodes. However we need to confirm that machine API is able to pick the remaining machines and provision the additional nodes in case we set a replica count > 3.
@pierreprinetti Could you share the results of your testing?

**Comment by pierreprinetti** [comment]
@mandre 
Tested with:
```
$ openshift-install version
openshift-install-linux-4.7.0-0.nightly-2021-01-10-070949 4.7.0-0.nightly-2021-01-10-070949
built from commit 69f0bbc18e8c6b1a6e278c54efa2def9b210033a
release image quay.io/openshift-release-dev/ocp-release-nightly@sha256:62768f2e8bb5850c469c1d15f6cd4337bd0f3f7ecb560b009c61e6d36a134fe8
```

The cluster correctly comes up with 5 Control plane nodes when `replicas` is set to 5 in `install-config`.

@staebler We are hardcoding 3 masters in Terraform, to work around a long-standing bad behaviour of OpenStack's Compute node. I believe it would be good to have a couple eyeballs from the Installer team on this.

File: data/data/openstack/masters/main.tf [hunk 2]
```diff
@@ -40,21 +40,84 @@ resource "openstack_compute_servergroup_v2" "master_group" {
...
+# when the servers are created concurrently.
+#
+# We chose to unroll the loop into three instances, because three is the
+# minimum number of required Control plane nodes, as stated in the
+# documentation[1].
+#
+# The expectation is that machine-api-operator will take care of creating any
```
Comment (staebler) [inline data/data/openstack/masters/main.tf:55]: The machine-api-operator is not involved in creating control plane nodes. Adding or removing control plane nodes after installation requires manual intervention by the user. However, there are plans to change this in the future.

**Review by staebler — COMMENTED** [review]
This is acceptable in the short term where there is a strict requirement for 3 master nodes. That requirement will be relaxed in a future release (maybe 4.9, 4.10). We need to come up with a better solution at some point.

File: data/data/openstack/masters/main.tf (hunk 2 above)
Comment (pierreprinetti) [reply data/data/openstack/masters/main.tf:55]: cf. https://issues.redhat.com/browse/OSASINFRA-2262

**Pull request merged**